RFCs/FS-1027-fsi-references.md

```diff
@@ -0,0 +1,100 @@
+* `#r "nuget: <package-name>, <package-version>"`
+
+This extends the "language" of `#r` to support implementation and reference assemblies, NuGet packages, Paket dependencies, and potentially more.
+
+# Motivation
+[motivation]: #motivation
+
+Motivation for this change is twofold:
+
+1. There is a [strong desire](https://github.com/fsharp/fslang-suggestions/issues/542) to reference packages via `#r` instead of assemblies.  There has also been [experimental work](https://github.com/Microsoft/visualfsharp/pull/2483) to support Paket in this way.
+2. .NET Core uses both Reference Assemblies and Implementation Assemblies, which FSI must understand.  It does not understand this split today.
+
+# Detailed design
+[design]: #detailed-design
+
+Supporting this design requires significant changes in how FSI references assemblies.  As stated above, .NET Core (and .NET Standard) introduces a new model by which assemblies are used and laid out on disk.  FSI cannot assume everything will be in the same place as it is in a .NET Framework world.
+
+To begin, existing behavior with `#r` will remain unchanged for FSI running in non-.NET Core.  That is, if you reference an assembly today on .NET Framework or Mono via `#r`, it will always continue to work in the same way that it always has.
+
+On .NET Core, however, `#r "<path-to-assembly>"` will be used for referencing *Reference Assemblies*.  To handle *Implementation Assemblies*, we introduce the notion of `#r impl: <path-to-implementation-assembly>`.  This is needed to directly reference implementation asseblies pulled down in packages or in the .NET Core Shared Framework.
+
+This split for .NET Core necessitates the introduction of `#r "nuget: name, version"` and `#r "paket: paket-command"` to simplify referencing dependencies in F# script files for .NET Core.  The intention is that FSI Scripts in the future will always specify dependencies via `#r "nuget` or `#r "paket` (or a future extended command, like `#r "project"`), allowing only FSI to be concerned with referencing the correct assemblies.  FSI in turn will defer to the specified package manager to resolve assemblies, load references in an ephemeral script, and load that script in the FSI session so that it has the references it needs.
+
+Here is an example:
+
+```fsharp
+#r "paket: nuget Newtonsoft.Json"
+
+open Newtonsoft.Json
+
+type Foo = { IntVal: int; StringVal: string }
+
+let foo = { IntVal = 12; StringVal = "bananas" }
+let json = JsonConvert.SerializeObject(foo)
+
+printfn "Object: %A\n JSON: %s" foo json
+```
+
+When FSI encounters this script, it will:
+
+1. Create an ephemeral script.
+2. Call into Paket to resolve dependencies.
+3. Generate a set of *Reference Assembly* and *Implementation Assembly* references in the ephemeral script, using `#r "path-to-ref-assembly"` and `#r "impl: path-to-impl-assembly""`
+4. Load the ephemeral script in the FSI session, thus giving FSI the correct references it needs to run in .NETCore
```

Comment:
Currently `#r-impl` or `#r "impl:` are the ones being considered, but that could just get hashed out in the issue.